core/src/program.rs

```diff
@@ -259,18 +259,72 @@ impl<EC: EvalCache> Program<EC> {
         String::from_utf8(buffer.into_inner().into_inner()).unwrap()
     }
 
+    /// Evaluate a program into a form suitable for extracting documentation.
+    /// This needs to evaluate to WHNF, but then still descend into subfields
+    /// whose values are records.
+    #[cfg(feature = "doc")]
+    pub fn eval_for_docs(&mut self) -> Result<RichTerm, Error> {
+        use crate::eval::{Closure, Environment};
+        use crate::match_sharedterm;
+        use crate::term::record::RecordData;
+        use crate::term::Term;
+
+        let (t, initial_env) = self.prepare_eval()?;
+
+        fn do_eval<EC: EvalCache>(
+            vm: &mut VirtualMachine<Cache, EC>,
+            t: RichTerm,
+            current_env: Environment,
+            initial_env: &Environment,
+        ) -> Result<RichTerm, Error> {
+            vm.reset();
+            let (rt, env) = vm
+                .eval_closure(
+                    Closure {
+                        body: t,
+                        env: current_env,
+                    },
+                    initial_env,
+                )
+                .map_err(Error::from)?;
+
+            match_sharedterm! {rt.term, with {
+                    Term::Record(data) => {
+                        let fields = data
+                            .fields
+                            .into_iter()
```

Comment:
I would have thought you would need to do some substitution at some point, but it turns out `eval_closure` is already evaluating variables away, mapping over fields is enough. Nice!